cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/ComputeEngineCredentials2.java

```diff
@@ -0,0 +1,154 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.runtime.spi.provisioner.dataproc;
+
+import com.google.api.client.util.ExponentialBackOff;
+import com.google.api.client.util.GenericData;
+import com.google.auth.oauth2.AccessToken;
+import com.google.auth.oauth2.GoogleCredentials;
+import com.google.common.io.CharStreams;
+import com.google.gson.Gson;
+import io.netty.handler.codec.http.HttpResponseStatus;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.io.Reader;
+import java.net.HttpURLConnection;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.util.Date;
+
+/**
+ * Fecth and refresh token
+ */
+public final class ComputeEngineCredentials2 extends GoogleCredentials {
+  private static final Logger LOG = LoggerFactory.getLogger(ComputeEngineCredentials2.class);
+  private static final Gson GSON = new Gson();
+  /**
+   * Time (in millisecond) to refresh the credentials before it expires.
+   */
+  private static final int NUMBER_OF_RETRIES = 10;
+  private static final int MIN_WAIT_TIME = 5000;
+  private static final int MAX_WAIT_TIME = 10000;
+  public static ComputeEngineCredentials2 computeEngineCredentials2;
+  private static byte[] lock = new byte[0];
+  private static GoogleCredentials cachedGoogleCredentials = null;
+  private final String endPoint;
+
+  private ComputeEngineCredentials2(String endPoint) {
+    this.endPoint = endPoint;
+  }
+
+  public static ComputeEngineCredentials2 build(String endPoint) {
+    if (computeEngineCredentials2 == null) {
+      synchronized (lock) {
+        if (computeEngineCredentials2 == null) {
+          computeEngineCredentials2 = new ComputeEngineCredentials2(endPoint);
+        }
+      }
+    }
+
+    try {
+      LOG.error("RETURNING computeEngineCredentials2 with " + computeEngineCredentials2.getAccessToken());
+    } catch (Exception ex) {
+      LOG.error("EXCEPTION", ex);
+    }
+    return computeEngineCredentials2;
+  }
+
+  /**
+   * Fetches and returns GoogleCredentials locally if endPoint is null, or remotely from the provided endPoint.
+   *
+   * @param endPoint endpoint to be used for fetching GoogleCredentials from.
+   * @return GoogleCredentials
+   * @throws IOException
+   */
+  public static AccessToken getComputeEngineCredentials(String endPoint) throws IOException {
+    if (endPoint != null) {
+      return getRemoteCredentials(endPoint);
+    }
+
+    return getLocalCredentials();
+  }
+
+  private static AccessToken getLocalCredentials() throws IOException {
+    LOG.error(">>>>> CALLING getLocalCredentials");
```

Comment:
This is a trace log? Also, the `>>>>` seems awkward.

Reply:
These were for testing purposes. They are gone now.